File: JDK11/opentraceability/src/main/java/opentraceability/mappers/epcis/xml/epciquerydocumentxmlmapper.java
package opentraceability.mappers.epcis.xml;

import opentraceability.models.events.EPCISQueryDocument;
import opentraceability.models.events.EPCISVersion;
import opentraceability.interfaces.IEvent;
import opentraceability.interfaces.IEPCISQueryDocumentMapper;
import opentraceability.mappers.OpenTraceabilityXmlMapper;
import opentraceability.utility.Constants;
import opentraceability.utility.OTLogger;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.IOException;
import java.io.StringWriter;
import java.util.Objects;

public class EPCISQueryDocumentXMLMapper implements IEPCISQueryDocumentMapper {

    @Override
    public EPCISQueryDocument map(String strValue, boolean checkSchema) throws Exception {
        try {
            var (document, xDoc) = EPCISDocumentBaseXMLMapper.readXml<EPCISQueryDocument>(strValue);

            if (xDoc.getDocumentElement() == null) {
                throw new Exception("Failed to parse EPCISQueryDocument from xml string because after parsing the XDocument the Root property was null.");
            }
            if (Objects.isNull(document.epcisVersion)) {
                throw new Exception("doc.epcisVersion is NULL. This must be set to a version.");
            }

            if (checkSchema) {
                EPCISDocumentBaseXMLMapper.validateEPCISQueryDocumentSchema(xDoc, Objects.requireNonNull(document.epcisVersion));
            }

            String epcisQueryXName = document.epcisVersion.equals(EPCISVersion.V1) ? Constants.EPCISQUERY_1_XNAMESPACE : Constants.EPCISQUERY_2_XNAMESPACE;

            Element xQueryName = xDoc.getDocumentElement().getFirstElementByXPath("EPCISBody/" + epcisQueryXName + ":QueryResults/queryName");

            if (xQueryName != null) {
                document.QueryName = xQueryName.getTextContent();
            }

            Element xEventList = xDoc.getDocumentElement().getFirstElementByXPath("EPCISBody/" + epcisQueryXName + ":QueryResults/resultsBody/EventList");

            if (xEventList != null) {
                for (int i = 0; i < xEventList.getChildNodes().getLength(); i++) {
                    Element xEvent = (Element) xEventList.getChildNodes().item(i);
                    if (xEvent != null) {
                        String eventType = EPCISDocumentBaseXMLMapper.getEventTypeFromProfile(xEvent);
                        IEvent e = OpenTraceabilityXmlMapper.fromXml(xEvent, eventType, Objects.requireNonNull(document.epcisVersion));
                        document.events.add(e);
                    }
                }
            }

            return document;
        } catch (Exception ex) {
            Exception exception = new Exception("Failed to parse the EPCIS document from the XML. xml= " + strValue, ex);
            OTLogger.error(exception);
            throw exception;
        }
    }

    @Override
    public String map(EPCISQueryDocument doc) throws Exception {

        if (Objects.isNull(doc.epcisVersion)) {
            throw new Exception("EPCISVersion is NULL. This must be set to a version.");
        }

        String epcisNS = doc.epcisVersion.equals(EPCISVersion.V2) ? Constants.EPCISQUERY_2_NAMESPACE : Constants.EPCISQUERY_1_NAMESPACE;

        var xDoc = EPCISDocumentBaseXMLMapper.writeXml(doc, epcisNS, "EPCISQueryDocument");

        if (xDoc.getDocumentElement() == null) {
            throw new Exception("Failed to parse EPCISQueryDocument from xml string because after parsing the XDocument the Root property was null.");
        }

        String epcisQueryXName = doc.epcisVersion.equals(EPCISVersion.V1) ? Constants.EPCISQUERY_1_XNAMESPACE : Constants.EPCISQUERY_2_XNAMESPACE;

        Element xQueryResults = xDoc.createElementNS(epcisQueryXName, "QueryResults");
        xQueryResults.appendChild(xDoc.createElement("queryName"));
        Element resultsBody = xDoc.createElement("resultsBody");
        resultsBody.appendChild(xDoc.createElement("EventList"));
        xQueryResults.appendChild(resultsBody);
        Objects.requireNonNull(xDoc.getDocumentElement().getElementsByTagName("EPCISBody")).item(0).appendChild(xQueryResults);

        Element xQueryName = xDoc.getDocumentElement().getFirstElementByXPath("EPCISBody/" + epcisQueryXName + ":QueryResults/queryName");
        if (xQueryName != null) {
            xQueryName.setTextContent(doc.QueryName);
        }

        Element xEventList = xDoc.getDocumentElement().getFirstElementByXPath("EPCISBody/" + epcisQueryXName + ":QueryResults/resultsBody/EventList");

        for (IEvent e : doc.events) {
            String xname = EPCISDocumentBaseXMLMapper.getEventXName(e);
            Element xEvent = OpenTraceabilityXmlMapper.toXml(xname, e, Objects.requireNonNull(doc.epcisVersion));
            if (Objects.requireNonNull(e.eventType).equals(opentraceability.models.events.EventType.TransformationEvent) && doc.epcisVersion == EPCISVersion.V1) {
                Element extension = xDoc.createElement("extension");
                extension.appendChild(xEvent);
                xEventList.appendChild(extension);
            } else {
                if (Objects.nonNull(xEvent)) {
                    xEventList.appendChild(xEvent);
                }
            }
        }

        EPCISDocumentBaseXMLMapper.validateEPCISQueryDocumentSchema(xDoc, Objects.requireNonNull(doc.epcisVersion));

        return docToString(xDoc);
    }

    private String docToString(org.w3c.dom.Document doc) throws Exception {
        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        javax.xml.transform.Transformer transformer;
        transformer = transformerFactory.newTransformer();
        StringWriter writer;
        writer = new StringWriter();
        transformer.transform(new DOMSource(doc), new StreamResult(writer));
        return writer.getBuffer().toString();
    }
}

File: Kotlin/OpenTraceability/src/main/kotlin/opentraceability/mappers/epcis/xml/epciquerydocumentxmlmapper.kt
package opentraceability.mappers.epcis.xml

import opentraceability.utility.*
import opentraceability.interfaces.IEvent
import opentraceability.interfaces.IEPCISQueryDocumentMapper
import opentraceability.mappers.OpenTraceabilityXmlMapper
import opentraceability.models.events.EPCISQueryDocument
import opentraceability.models.events.EPCISVersion
import org.w3c.dom.Element
import org.w3c.dom.NodeList
import java.io.StringWriter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

class EPCISQueryDocumentXMLMapper : IEPCISQueryDocumentMapper {
    override fun map(strValue: String, checkSchema: Boolean): EPCISQueryDocument {
        try {
            var (document, xDoc) = EPCISDocumentBaseXMLMapper.readXml<EPCISQueryDocument>(strValue)

            if (xDoc.documentElement == null) {
                throw Exception("Failed to parse EPCISQueryDocument from xml string because after parsing the XDocument the Root property was null.")
            }
            if (document.epcisVersion == null) {
                throw Exception("doc.epcisVersion is NULL. This must be set to a version.")
            }

            if (checkSchema) {
                EPCISDocumentBaseXMLMapper.validateEPCISQueryDocumentSchema(xDoc, document.epcisVersion!!)
            }

            val epcisQueryXName =
                if (document.epcisVersion == EPCISVersion.V1) opentraceability.Constants.EPCISQUERY_1_XNAMESPACE else opentraceability.Constants.EPCISQUERY_2_XNAMESPACE

            val xQueryName =
                xDoc.documentElement?.getFirstElementByXPath("EPCISBody/$epcisQueryXName:QueryResults/queryName")
            if (xQueryName != null) {
                document.QueryName = xQueryName.nodeValue
            }

            val xEventList =
                xDoc.documentElement?.getFirstElementByXPath("EPCISBody/$epcisQueryXName:QueryResults/resultsBody/EventList")
                    ?: throw Exception("Failed to get EPCISBody/EventList after adding it to the XDoc.Root")

            if (xEventList != null) {
                for (i in 0 until xEventList.childNodes.length) {
                    val xEvent = xEventList.childNodes.item(i) as? Element
                    if (xEvent != null) {
                        val eventType = EPCISDocumentBaseXMLMapper.getEventTypeFromProfile(xEvent)
                        val e = OpenTraceabilityXmlMapper.fromXml(xEvent, eventType, document.epcisVersion!!)
                        document.events.add(e as IEvent)
                    }
                }
            }

            return document
        } catch (ex: Exception) {
            val exception = Exception("Failed to parse the EPCIS document from the XML. xml=$strValue", ex)
            opentraceability.OTLogger.error(exception)
            throw exception
        }
    }

    override fun map(doc: EPCISQueryDocument): String {
        if (doc.epcisVersion == null) {
            throw Exception("EPCISVersion is NULL. This must be set to a version.")
        }

        val epcisNS =
            if (doc.epcisVersion == EPCISVersion.V2) opentraceability.Constants.EPCISQUERY_2_NAMESPACE else opentraceability.Constants.EPCISQUERY_1_NAMESPACE

        val xDoc = EPCISDocumentBaseXMLMapper.writeXml(doc, epcisNS, "EPCISQueryDocument")
        if (xDoc.documentElement == null) {
            throw Exception("Failed to parse EPCISQueryDocument from xml string because after parsing the XDocument the Root property was null.")
        }

        val epcisQueryXName =
            if (doc.epcisVersion == opentraceability.models.events.EPCISVersion.V1) opentraceability.Constants.EPCISQUERY_1_XNAMESPACE else opentraceability.Constants.EPCISQUERY_2_XNAMESPACE

        // write the query name
        val xQueryResults = xDoc.createElementNS(epcisQueryXName, "QueryResults")
        xQueryResults.appendChild(xDoc.createElement("queryName"))
        val resultsBody = xDoc.createElement("resultsBody")
        resultsBody.appendChild(xDoc.createElement("EventList"))
        xQueryResults.appendChild(resultsBody)
        xDoc.documentElement?.getElementsByTagName("EPCISBody")?.item(0)?.appendChild(xQueryResults)

        val xQueryName =
            xDoc.documentElement?.getFirstElementByXPath("EPCISBody/$epcisQueryXName:QueryResults/queryName");
        if (xQueryName != null) {
            xQueryName.nodeValue = doc.QueryName
        }

        // write the events
        val xEventList =
            xDoc.documentElement?.getFirstElementByXPath("EPCISBody/$epcisQueryXName:QueryResults/resultsBody/EventList")
                ?: throw Exception("Failed to get EPCISBody/EventList after adding it to the XDoc.Root")
        for (e in doc.events) {
            val xname = EPCISDocumentBaseXMLMapper.getEventXName(e)
            val xEvent = OpenTraceabilityXmlMapper.toXml(xname, e, doc.epcisVersion!!)
            if (e.eventType == opentraceability.models.events.EventType.TransformationEvent && doc.epcisVersion == EPCISVersion.V1) {
                val extension = xDoc.createElement("extension")
                extension.appendChild(xEvent)
                xEventList.appendChild(extension)
            } else {
                xEvent?.let { xEventList.appendChild(it) }
            }
        }

        EPCISDocumentBaseXMLMapper.validateEPCISQueryDocumentSchema(xDoc, doc.epcisVersion!!)

        return docToString(xDoc)
    }

    fun docToString(doc: org.w3c.dom.Document): String {
        val transformerFactory = TransformerFactory.newInstance()
        val transformer = transformerFactory.newTransformer()
        val writer = StringWriter()
        transformer.transform(DOMSource(doc), StreamResult(writer))
        return writer.toString()
    }
}
